File: funcscript/funcs/keyvalue/KvcNoneNullMemberFunction.cs
using funcscript.core;
using funcscript.model;
using System;

namespace funcscript.funcs.keyvalue
{
    public class KvcNoneNullMemberFunction : IFsFunction
    {
        public int MaxParsCount => 2;

        public CallType CallType => CallType.Infix;

        public string Symbol => "?.";

        public int Precidence => 200;

        private object EvaluateInternal(object target, object key)
        {
            if (!(key is string))
                throw new error.TypeMismatchError($"{Symbol} function: The second parameter should be a string (Member key).");

            if (target == null)
                return null;

            if (!(target is KeyValueCollection))
                throw new error.TypeMismatchError($"{Symbol} function: Cannot access member '{key}' on non-KeyValueCollection type '{FuncScript.GetFsDataType(target)}'.");

            return ((KeyValueCollection)target).Get(((string)key).ToLower());
        }

        public object Evaluate(IFsDataProvider parent, IParameterList pars)
        {
            if (pars.Count != MaxParsCount)
                throw new error.TypeMismatchError($"{Symbol} function: Expected {MaxParsCount} parameters, received {pars.Count}.");

            
            var key = pars.GetParameter(parent,1);
            var target = pars.GetParameter(parent,0);
            

            return EvaluateInternal(target, key);
        }

        public string ParName(int index)
        {
            return index switch
            {
                0 => "Key-value collection",
                1 => "Member key",
                _ => string.Empty,
            };
        }

    }
}

File: js_port/fs_js/src/funcs/keyvalue/KvcNoneNullMemberFunction.js
const { BaseFunction, CallType } = require('../../core/functionBase');
const helpers = require('../helpers');
const { requireString } = helpers;

class KvcNoneNullMemberFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = '?.';
    this.callType = CallType.Infix;
    this.precidence = 200;
  }

  get maxParameters() {
    return 2;
  }

  evaluate(provider, parameters) {
    const error = helpers.expectParamCount(this.symbol, parameters, this.maxParameters);
    if (error) {
      return error;
    }

    const target = parameters.getParameter(provider, 0);
    if (helpers.typeOf(target) === helpers.FSDataType.Null) {
      return helpers.typedNull();
    }

    const keyResult = requireString(this.symbol, parameters.getParameter(provider, 1), 'Member key');
    if (!keyResult.ok) {
      return keyResult.error;
    }

    const typedTarget = helpers.ensureTyped(target);
    if (helpers.typeOf(typedTarget) !== helpers.FSDataType.KeyValueCollection) {
      return helpers.makeError(
        helpers.FsError.ERROR_TYPE_MISMATCH,
        `${this.symbol} function: Cannot access member '${keyResult.value}' on type ${helpers.typeOf(typedTarget)}`
      );
    }

    const collection = helpers.valueOf(typedTarget);
    return collection.get(keyResult.value.toLowerCase());
  }
}

module.exports = {
  KvcNoneNullMemberFunction
};
